File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-9.cpp
class Solution {
public:
    unordered_map<int, unordered_map<int, int>> dp;
    vector<int> prices_;
    int maxProfit(vector<int>& prices) {
        prices_ = prices;
        return dfs(0, 1);
    }
    int dfs(int i, int canbuy) {
        if (i >= prices_.size()) {return 0;}
        if (dp.count(i) and dp[i].count(canbuy)) {
            return dp[i][canbuy];
        }
        int cooldown = dfs(i + 1, canbuy);
        int profit;
        // 1
        if (canbuy) {
            profit = dfs(i + 1, 0) - prices_[i];
        } else {
            profit = dfs(i + 2, 1) + prices_[i];
        }
        dp[i][canbuy] = max(cooldown, profit);
        return dp[i][canbuy];
    }
};

File: Data Structures & Algorithms/word-break/submission-9.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = {len(s):True}
        def dfs(i: int) -> bool:
            if i in dp:
                return dp[i]
            for w in wordDict:
                j = len(w)
                if i + j <= len(s) and s[i:i + j] == w:
                    if dfs(i + j):
                        dp[i] = True
                        return True
            dp[i] = False
            return False
        return dfs(0)
